KAFKA-5532: Making bootstrap.servers property a first citizen option for the ProducerPerformance

ppatierno opened the pull request:
Added bootstrap-server as option for specifying Kafka cluster to connect to

ppatierno committed:
Added bootstrap-server as option for specifying Kafka cluster to connect to

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/5766/
Test PASSed (JDK 7 and Scala 2.11).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/5752/
Test PASSed (JDK 8 and Scala 2.12).

ppatierno commented:
@ijuma what do you think about that ?

ijuma commented on tools/src/main/java/org/apache/kafka/tools/ProducerPerformance.java:
```diff
@@ -198,6 +200,13 @@ private static ArgumentParser argParser() {
 
+        parser.addArgument("--bootstrap-server")
+                .action(store())
+                .required(true)
```
We can't make it required because that is not backwards compatible.

ppatierno replied:
@ijuma Done. As a newbie on Kafka source code and policies I'm learning this kind of stuff. Could it be worth then open a KIP for making it required in order to have the `--bootstrap-server` option a really first citizen here ?

ppatierno committed:
Removed "required" for the new --bootstrap-server option for backwards compatibility

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6558/
Test PASSed (JDK 8 and Scala 2.12).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6573/
Test PASSed (JDK 7 and Scala 2.11).

ppatierno commented:
@ijuma do you think that it's ok now ?

ijuma commented:
Can we add a simple unit test?

ppatierno commented:
@ijuma I agree but in this case I need to startup an "embedded" Kafka cluster before launching the main. I see that something like that is provided by the `EmbeddedKafkaCluster` class but it is in the streams module and of course there is no dependency with the tools module.
I see that the tests for the tools in the core module extends the `KafkaServerTestHarness` class so something like the embedded cluster as before.
Some other classes like the `ConsoleConsumer` are more testable because for example the `ConsoleConsumer` expose the `ConsumerConfig` method useful for verifying args and `process` for executing the action.
What about the `ProducerPerformance` case ?

ijuma commented:
We could have a method that returns the `Properties` used by the producer. The unit test would simply check that the property was populated correctly. No need for starting the Kafka broker.

omkreddy commented:
+1

ppatierno commented:
@ijuma this kind of simple test is fine for me but it means that I have to split the `ProducerPerformance` logic because just calling the `main` method will start the producer (other than set the configuration) so the cluster is needed. I'm going to split the logic in a way that we can call a "configuration" and a "run" method from the unit test but they will be called inside the main method so that nothing changes from a user and kafka-producer-perf-test.sh perspective.

ijuma commented:
Yes, sure. That kind of refactoring is good as it makes the code more testable.

ppatierno committed:
Refactoring ProducerPerformance for more testability
Added CommandOptions base class for parsing arguments for command line tools
Added CommandLineUtils with some command line utilities
Added ProducerPerformanceOptions extending CommandOptions for ProducerPerformance tool class
Added unit test for new bootstrap-server argument

ppatierno commented:
@ijuma sorry for the delay on committing the refactored version for having code more testable but I thought about it and it turned out (for me) that this PR could be a good way for starting to add the `CommandOptions` base class I have already introduced in the WIP #3514 (on `TopicCommand` work) which other developers are using/waiting for a trunk merged version in order to use it for other tools refactoring (as [KAFKA-3268](https://issues.apache.org/jira/browse/KAFKA-3268)).

ppatierno committed:
Added new unit test on overridding bootstrap server using producer-props option

adyach commented on tools/src/main/java/org/apache/kafka/tools/CommandOptions.java:
```diff
@@ -0,0 +1,80 @@
+        if (this.ns.get(option) instanceof Boolean)
+            return this.ns.getBoolean(option);
+        else
+            return this.ns.getAttrs().get(option) != null;
```
the same will be `ns.get(option)`

ppatierno replied:
@adyach yes you are right ;)

adyach commented on tools/src/main/java/org/apache/kafka/tools/CommandOptions.java:
```diff
@@ -0,0 +1,80 @@
+     */
+    public boolean has(String option) {
+
+        if (this.ns.get(option) instanceof Boolean)
```
why is it so specific for Boolean type?

ppatierno replied:
@adyach from my experience developing the `TopicCommand` the reason for this specific type checking is the following.
The kafka-topics.sh tool (which use `TopicCommand`) has some options like `--create`, `--describe` and so on which doesn't have a related argument/value; they are a sort of "boolean" and you say that you want "to create" (true) but not "to describe" (false).
The way to have that is using `Arguments.storeTrue()` method.
For example : 

`
operationOptions.addArgument("--create")
                    .action(storeTrue())
                    .dest(CREATE)
                    .help("Create a new topic.");
`

If you don't use the `storeTrue()` method for such options without arg, an `ArgumentParserException` exception is raised.
Said that, when the parsing is executed, in the namespace attributes list you will have "create" entry with value true and "describe" entry with value false (even if you have't passed `--describe` on the command line).
Having just the line : 

`this.ns.get(option) != null`

means that for example in my `TopicCommand` case, for both "create" and "describe" this return true because they are in the attributes list (but "create" is true while "describe" is false).
To the command class it looks like the user has specified both `--create` and `--describe` which is not possible ... so an error is raised.
For this reason for `Boolean` values I need to check the value entry and not just checking that the entry is or not in the attributes list.

adyach replied:
Okay, it is clear now, thank you 👍 
However, I do not see you, using this method in the code. Instead, you just do e.g. 
`String bootstrapServer = opts.bootstrapServer();`
and then 
`if (bootstrapServer != null) {`
So you need a value in case it is not `null`, probably, that's why the `has` method won't be very usable. 
I am not against it, it adds no complexity, I am just not able to find a use case for it, except checking for flags, which can be checked by client code without any type checking.

ppatierno replied:
If you take a look at my WIP PR (https://github.com/apache/kafka/pull/3514) about the `TopicCommand` you can see that there are some other `CommandLineUtils` methods for checking arguments. The `has` method is used there and it provides a more concise way to check option existence. If you take a look at what is used in the current `TopicCommand` in Scala, the `has` method is provided by the `OptionSet` class (i.e. https://github.com/apache/kafka/blob/trunk/core/src/main/scala/kafka/utils/CommandLineUtils.scala#L33). I'd like to give quite the same developer experience but in Java so the decision to add this method to the `CommandOptions` base class.

viktorsomogyi replied:
I can agree with using the has() method, it proved useful in my commands too.

ppatierno committed:
Fixed way to access Namespace underlying attributes list

viktorsomogyi commented on tools/src/main/java/org/apache/kafka/tools/CommandOptions.java:
```diff
@@ -0,0 +1,80 @@
+
+        this.prepareArgs();
+
+        if (args.length == 0) {
```
I don't really understand this line. Wouldn't parse args check if there are no arguments since you have required args?

viktorsomogyi commented on tools/src/main/java/org/apache/kafka/tools/ProducerPerformance.java:
```diff
@@ -16,67 +16,102 @@
+     * @return  properties needed for configuring the used Kafka producer
+     * @throws Exception
+     */
+    Properties getProducerProps(ProducerPerformanceOptions opts) throws Exception {
```
Can you please add a comment that this method is visible only because of tests? (I suppose that was the idea as it is being used in the tests apart from this class from what I can see.)

viktorsomogyi commented on tools/src/main/java/org/apache/kafka/tools/ProducerPerformance.java:
```diff
@@ -159,140 +180,205 @@ public static void main(String[] args) throws Exception {
-        }
 
+            opts.parser.handleError(e);
+            Exit.exit(1);
```
Are these above two lines would be somewhat equivalent with printUsageAndDie? What is the difference?

ppatierno replied:
It's more like `printErrorAndDie`. I'm going to change with this. Thanks !

viktorsomogyi commented on tools/src/main/java/org/apache/kafka/tools/ProducerPerformance.java:
```diff
@@ -159,140 +180,205 @@ public static void main(String[] args) throws Exception {
+
+            payloadOptions.addArgument("--record-size")
+                    .action(store())
+                    .required(false)
```
Should we specify explicitly if this is the default?

ppatierno replied:
It's already there in the current (trunk) status of this tool. I prefer to leave it even for clarity (who don't know what the default is)

viktorsomogyi replied:
I see, ok :)

viktorsomogyi commented on tools/src/test/java/org/apache/kafka/tools/ProducerPerformanceTest.java:
```diff
@@ -0,0 +1,75 @@
+            assertEquals(props.get(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG), "server:9092");
+
+        } catch (Exception e) {
+            e.printStackTrace();
```
Generally I prefer using rules, makes your tests clearer. Try it if you like it :) http://junit.org/junit4/javadoc/4.12/org/junit/rules/ExpectedException.html
```
 @Rule
 public ExpectedException thrown= ExpectedException.none();
 
 @Test
 public void throwsException() {
     thrown.expect(NullPointerException.class);
     thrown.expectMessage("happened");
     throw new NullPointerException("What happened?");
 }
```

ppatierno replied:
Well, the try/catch is there just because the `getProducerProps` can throw an exception. I'm going to remove the try/catch having the test method marked with `throws Exception`. Btw, I like rules as well ;)

viktorsomogyi reviewed (commented):
Looked at your PR and made some minor comments. I'm happy that we are refactoring the command line tools, it is beneficial to the community.

ppatierno commented:
@ijuma related to the 1.0.0 release, this PR add a new flag but it's a refactoring as well. Do you think we can get it into 1.0.0 with a review ? It adds some classes useful for future use of Java for tools, so maybe it could be good having them from this release. Wdyt ?

ppatierno committed:
Removed try/catch on tests
Used printErrorAndDie for exiting on error

ppatierno commented:
@ijuma I got some feedback from @viktorsomogyi and applied in some way his suggestion. Maybe you could have a review on that for going ahead ? Thanks !

ppatierno committed:
Removed unused imports

ijuma reviewed (commented):
This PR includes a number of files in the `out` directory that should not be there.

ppatierno committed:
Removed out directory files

asfgit commented:
FAILURE
 No test results found.
--none--

asfgit commented:
FAILURE
 5891 tests run, 0 skipped, 23 failed.
--none--

asfgit commented:
FAILURE
 5891 tests run, 0 skipped, 23 failed.
--none--

ppatierno commented:
@ijuma just removed sorry.

ppatierno committed:
Fixed doubled word in a parameter description

asfgit commented:
FAILURE
 No test results found.
--none--

asfgit commented:
FAILURE
 5891 tests run, 0 skipped, 23 failed.
--none--

asfgit commented:
FAILURE
 5892 tests run, 0 skipped, 1 failed.
--none--

ppatierno committed:
Merge branch 'trunk' into kafka-5532

asfgit commented:
FAILURE
 4021 tests run, 1 skipped, 1 failed.
--none--

asfgit commented:
SUCCESS 
 8073 tests run, 5 skipped, 0 failed.
--none--

asfgit commented:
SUCCESS 
 8073 tests run, 5 skipped, 0 failed.
--none--

ppatierno commented:
@ijuma no other updates/requests on this PR after 15 days ?
